### ISSUE TITLE
Add API to lint AST instead of source code

### ISSUE THREAD

**Issue by bartlomieju** [denoland/deno_lint#580]
For the purpose of https://github.com/denoland/deno/issues/8641 we should have additional APIs on `Linter` struct that allows to pass already parsed AST instead of source code.

Current API:
https://github.com/denoland/deno_lint/blob/14a4dccfbd84e13f3fc2e245961f3c84f3ffc339/src/linter.rs#L212-L216

Proposed API:
```rust
  pub fn lint_ast(
    &mut self,
    file_name: String,
    ast: swc_ecmascript::ast::Program,
  ) -> Result<
    (Rc<swc_common::SourceFile>, Vec<LintDiagnostic>),
    SwcDiagnosticBuffer,
  >
```

**Comment by kitsonk** [comment]
small nit... `lint_with_ast`?

Are you proposing maintaining existing API as well?

**Comment by bartlomieju** [comment]
> small nit... `lint_with_ast`?

Sounds good to me.

> Are you proposing maintaining existing API as well?

Yes, most of the code would still be shared, if it proves to difficult in the long run I'll be happy to remove existing APIs.

**Comment by magurotuna** [comment]
I'll work on it.
